cloud/edgecontroller/pkg/devicecontroller/controller/downstream.go

```diff
@@ -0,0 +1,742 @@
+			checkModelExists = true
+			break
+		}
+	}
+	if checkModelExists != true {
+		addDeviceModelAndVisitors(deviceModel, deviceProfile)
+	}
+	bytes, err := json.Marshal(deviceProfile)
+	if err != nil {
+		log.LOGGER.Errorf("Failed to marshal deviceprofile: %v", deviceProfile)
+		return
+	}
+	configMap.Data["deviceProfile.json"] = string(bytes)
+}
+
+func addDeviceModelAndVisitors(deviceModel *CacheDeviceModel, deviceProfile *types.DeviceProfile) {
+	model := &types.DeviceModel{}
+	model.Name = deviceModel.Name
+	model.Properties = make([]*types.Property, 0)
+	for _, ppt := range deviceModel.Spec.Properties {
+		property := &types.Property{}
+		property.Name = ppt.Name
+		property.Description = ppt.Description
+		if ppt.Type.Int != nil {
+			property.AccessMode = string(ppt.Type.Int.AccessMode)
+			property.DataType = DataTypeInt
+			property.DefaultValue = ppt.Type.Int.DefaultValue
+			property.Maximum = ppt.Type.Int.Maximum
+			property.Minimum = ppt.Type.Int.Minimum
+			property.Unit = ppt.Type.Int.Unit
+		} else if ppt.Type.String != nil {
+			property.AccessMode = string(ppt.Type.String.AccessMode)
+			property.DataType = DataTypeString
+			property.DefaultValue = ppt.Type.String.DefaultValue
+		}
+		model.Properties = append(model.Properties, property)
+	}
+	deviceProfile.DeviceModels = append(deviceProfile.DeviceModels, model)
+	for _, pptv := range deviceModel.Spec.PropertyVisitors {
+		propertyVisitor := &types.PropertyVisitor{}
+		propertyVisitor.Name = pptv.PropertyName
+		propertyVisitor.PropertyName = pptv.PropertyName
+		propertyVisitor.ModelName = deviceModel.Name
+		if pptv.Modbus != nil {
+			propertyVisitor.Protocol = Modbus
+			propertyVisitor.VisitorConfig = pptv.Modbus
+		} else if pptv.OpcUA != nil {
+			propertyVisitor.Protocol = OPCUA
+			propertyVisitor.VisitorConfig = pptv.OpcUA
+		} else if pptv.Bluetooth != nil {
+			propertyVisitor.Protocol = Bluetooth
+			propertyVisitor.VisitorConfig = pptv.Bluetooth
+		}
+		deviceProfile.PropertyVisitors = append(deviceProfile.PropertyVisitors, propertyVisitor)
+	}
+}
+
+func addDeviceInstanceAndProtocol(device *v1alpha1.Device, deviceProfile *types.DeviceProfile) {
+	deviceInstance := &types.DeviceInstance{}
+	deviceProtocol := &types.Protocol{}
+	deviceInstance.ID = device.Name
+	deviceInstance.Name = device.Name
+	deviceInstance.Model = device.Spec.DeviceModelRef.Name
+	var protocol string
+	if device.Spec.Protocol.OpcUA != nil {
+		protocol = OPCUA + "-" + device.Name
+		deviceInstance.Protocol = protocol
+		deviceProtocol.Name = protocol
+		deviceProtocol.Protocol = OPCUA
+		deviceProtocol.ProtocolConfig = device.Spec.Protocol.OpcUA
+	} else if device.Spec.Protocol.Modbus.RTU != nil {
+		protocol = ModbusRTU + "-" + device.Name
+		deviceInstance.Protocol = protocol
+		deviceProtocol.Name = protocol
+		deviceProtocol.Protocol = ModbusRTU
+		deviceProtocol.ProtocolConfig = device.Spec.Protocol.Modbus.RTU
+	} else if device.Spec.Protocol.Modbus.TCP != nil {
```

Comment:
we need to add for bluetooth as well.

Reply:
bluetooth currently doesn't have visitor config defined.